packages/reakit/src/Step/StepContainer.tsx

```diff
@@ -24,230 +18,161 @@ export interface StepContainerState {
-  state.ids[state.current];
-
-export const hasPrevious = () => (state: StepContainerState) =>
-  state.ids.length > 1 && !!state.ids[state.current - 1];
-
-export const hasNext = () => (state: StepContainerState) =>
-  state.ids.length > 1 && !!state.ids[state.current + 1];
-
-export const indexOf = (idOrIndex: string | number) => (
-  state: StepContainerState
-) => (typeof idOrIndex === "number" ? idOrIndex : state.ids.indexOf(idOrIndex));
+const selectors: SelectorMap<StepContainerState, StepContainerSelectors> = {
+  getCurrentId: () => state => state.ids[state.current],
 
-export const isCurrent = (idOrIndex: string | number) => (
-  state: StepContainerState
-) => state.current >= 0 && state.current === indexOf(idOrIndex)(state);
+  hasPrevious: () => state =>
+    state.ids.length > 1 && !!state.ids[state.current - 1],
 
-// actions
-export const show = (idOrIndex: string | number) => (
-  state: StepContainerState
-) => ({
-  current: indexOf(idOrIndex)(state)
-});
+  hasNext: () => state =>
+    state.ids.length > 1 && !!state.ids[state.current + 1],
 
-export const hide = () => () => ({ current: -1 });
+  indexOf: (idOrIndex: string | number) => state =>
+    typeof idOrIndex === "number" ? idOrIndex : state.ids.indexOf(idOrIndex),
 
-export const toggle = (idOrIndex: string | number) => (
-  state: StepContainerState
-) => (isCurrent(idOrIndex)(state) ? hide()() : show(idOrIndex)(state));
-
-export const previous = () => (state: StepContainerState) => {
-  if (hasPrevious()(state)) {
-    return { current: state.current - 1 };
-  }
-  if (state.loop) {
-    return { current: state.ids.length - 1 };
-  }
-  return {};
+  isCurrent: (idOrIndex: string | number) => state =>
+    state.current >= 0 && state.current === selectors.indexOf(idOrIndex)(state)
 };
 
-export const next = () => (state: StepContainerState) => {
-  if (hasNext()(state)) {
-    return { current: state.current + 1 };
-  }
-  if (state.loop) {
-    return { current: 0 };
-  }
-  return {};
-};
+const actions: ActionMap<StepContainerState, StepContainerActions> = {
+  show: (idOrIndex: string | number) => state => ({
+    current: selectors.indexOf(idOrIndex)(state)
+  }),
 
-export const reorder = (id: string, order = 0) => (
-  state: StepContainerState
-) => {
-  const ordered = { ...state.ordered, [id]: order };
-  const ids = state.ids
-    .slice()
-    .sort((a, b) => (ordered[a] || 0) - (ordered[b] || 0));
-  return {
-    ordered,
-    ids,
-    ...(isCurrent(id)(state) ? show(id)({ ...state, ids }) : {})
-  };
-};
+  hide: () => ({ current: -1 }),
 
-export const register = (id: string, order = 0) => (
-  state: StepContainerState
-) => {
-  const ids = state.ids.indexOf(id) >= 0 ? state.ids : [...state.ids, id];
-  return reorder(id, order)({ ...state, ids });
-};
+  toggle: (idOrIndex: string | number) => state =>
+    selectors.isCurrent(idOrIndex)(state)
+      ? callMeMaybe(actions.hide(), state)
```

Comment:
Actions are functions that might return whether another function or an object. This is the same as React `setState` first parameter.

That `callMeMaybe` handles that.